pkg/provider/barbican/provider.go

```diff
@@ -0,0 +1,194 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package barbican
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+)
+
+// Provider is a secrets provider for OpenStack Barbican.
+// It implements the necessary NewClient() and ValidateStore() funcs.
+type Provider struct{}
+
+var _ esv1beta1.Provider = &Provider{}
+
+const (
+	errInitProvider        = "unable to initialize barbican provider: %s"
+	errNilStore            = "found nil store"
+	errMissingStoreSpec    = "store is missing spec"
+	errMissingProvider     = "storeSpec is missing provider"
+	errInvalidProviderSpec = "invalid provider spec. Missing Barbican field in store %s"
+)
+
+func init() {
+	esv1beta1.Register(&Provider{}, &esv1beta1.SecretStoreProvider{
+		Barbican: &esv1beta1.BarbicanProvider{},
+	})
+}
+
+var useMu = sync.Mutex{}
+
+func (p *Provider) Capabilities() esv1beta1.SecretStoreCapabilities {
+	return esv1beta1.SecretStoreReadWrite
+}
+
+func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore, kube kclient.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	return newClient(ctx, store, kube, namespace)
+}
+
+func newClient(ctx context.Context, store esv1beta1.GenericStore, kube kclient.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	if store == nil {
+		return nil, fmt.Errorf(errInitProvider, "nil store")
+	}
+	storeSpec := store.GetSpec()
+	if storeSpec == nil || storeSpec.Provider == nil || storeSpec.Provider.Barbican == nil {
+		return nil, fmt.Errorf(errBarbicanStore)
+	}
+	bStore := storeSpec.Provider.Barbican
+
+	password, err := getPasswordFromSecrets(ctx, storeSpec.Provider.Barbican.Auth, kube, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	if storeSpec.Provider.Barbican.ServiceName == "" {
+		storeSpec.Provider.Barbican.ServiceName = "barbican"
+	}
+
+	useMu.Lock()
+	defer useMu.Unlock()
+
+	authOpts := gophercloud.AuthOptions{
+		IdentityEndpoint: storeSpec.Provider.Barbican.AuthUrl,
+		Username:         storeSpec.Provider.Barbican.Username,
+		Password:         password,
+		DomainID:         storeSpec.Provider.Barbican.UserDomain,
+		TenantName:       storeSpec.Provider.Barbican.ProjectName,
+	}
```

Comment:
Hey, can you please also expose `ApplicationCredential(ID|Name|Secret)`, maybe also `Passcode` here? That would help users a lot!

Reply:
Nice. Thank you for the review. I'm happy that it works. Sure, I will add an additional auth method with ApplicationCredential.

Reply:
> Hey, can you please also expose `ApplicationCredential(ID|Name|Secret)`, maybe also `Passcode` here? That would help users a lot!

Hey. I already exposed ApplicationCredential. You can test it